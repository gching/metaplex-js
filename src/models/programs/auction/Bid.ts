import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { BorshSchema } from '../../utils/types';

export class Bid {
  static SCHEMA: BorshSchema = new Map([
    [
      Bid,
      {
        kind: 'struct',
        fields: [
          ['key', 'pubkey'],
          ['amount', 'u64'],
        ],
      },
    ],
  ]);

  key: PublicKey;
  amount: BN;
  constructor(args: { key: PublicKey; amount: BN }) {
    this.key = args.key;
    this.amount = args.amount;
  }
}
