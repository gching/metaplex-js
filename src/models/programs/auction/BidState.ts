import BN from 'bn.js';
import { PublicKey } from '@solana/web3.js';
import { Bid } from './Bid';
import { BorshSchema } from '../../utils/types';

export enum BidStateType {
  EnglishAuction = 0,
  OpenEdition = 1,
}

export class BidState {
  static SCHEMA: BorshSchema = new Map([
    [
      BidState,
      {
        kind: 'struct',
        fields: [
          ['type', 'u8'],
          ['bids', [Bid]],
          ['max', 'u64'],
        ],
      },
    ],
    ...Bid.SCHEMA,
  ]);

  type: BidStateType;
  bids: Bid[];
  max: BN;

  public getWinnerAt(winnerIndex: number): PublicKey | null {
    const convertedIndex = this.bids.length - winnerIndex - 1;

    if (convertedIndex >= 0 && convertedIndex < this.bids.length) {
      return this.bids[convertedIndex].key;
    } else {
      return null;
    }
  }

  public getWinnerIndex(bidder: PublicKey): number | null {
    if (!this.bids) return null;

    const index = this.bids.findIndex(
      (b) => b.key.toBase58() === bidder.toBase58(),
    );
    // auction stores data in reverse order
    if (index !== -1) {
      const zeroBased = this.bids.length - index - 1;
      return zeroBased < this.max.toNumber() ? zeroBased : null;
    } else return null;
  }

  constructor(args: { type: BidStateType; bids: Bid[]; max: BN }) {
    this.type = args.type;
    this.bids = args.bids;
    this.max = args.max;
  }
}
