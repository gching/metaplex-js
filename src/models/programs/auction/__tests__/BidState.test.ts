import BN from 'bn.js';
import { BidState, BidStateType } from '../BidState';
import { Bid } from '../Bid';
import { PublicKey } from '@solana/web3.js';

describe('BidState', () => {
  it('should instantiate with the correct values', () => {
    const bidState = new BidState({
      bids: [
        new Bid({
          key: PublicKey.default,
          amount: new BN(0),
        }),
      ],
      max: new BN(0),
      type: BidStateType.EnglishAuction,
    });

    expect(bidState).toMatchSnapshot();
  });
});
