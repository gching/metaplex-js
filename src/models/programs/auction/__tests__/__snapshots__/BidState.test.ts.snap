// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BidState should instantiate with the correct values 1`] = `
BidState {
  "bids": Array [
    Bid {
      "amount": "00",
      "key": PublicKey {
        "_bn": "00",
      },
    },
  ],
  "max": "00",
  "type": 0,
}
`;
